File: Laravel/app/Http/Controllers/PortalController.php
<?php

namespace App\Http\Controllers;

use App\Bd\Auth;
use App\Bd\Bd;
use App\Bd\Client;
use Illuminate\Http\RedirectResponse;
use Illuminate\Http\Request;
use Illuminate\View\View;

/**
 * Customer portal. Everything here is per-customer, so nothing is cached —
 * `Bd::attempt()` rather than `Bd::remember()`.
 */
class PortalController extends Controller
{
    public function index(Request $request): View|RedirectResponse
    {
        $session = $this->session($request);
        if (! $session) {
            return redirect('/api/bd-auth/sign-in?returnTo=/my-account');
        }

        $token = $request->cookie(config('bd.session_cookie'));

        $work = Bd::attempt(
            static fn (Client $c) => $c->portal($token, $session['organizationId'] ?? null)->getWork(),
            default: null,
        );

        return view('pages.my-account', [
            'user' => $session['user'] ?? null,
            'work' => $work,
        ]);
    }

    public function submitReview(Request $request): RedirectResponse
    {
        $validated = $request->validate([
            'rating' => ['required', 'integer', 'min:1', 'max:5'],
            'body' => ['required', 'string', 'max:4000'],
        ]);

        $session = $this->session($request);
        if (! $session) {
            return redirect('/api/bd-auth/sign-in?returnTo=/my-account');
        }

        $token = $request->cookie(config('bd.session_cookie'));

        $result = Bd::attempt(static fn (Client $c) => $c
            ->portal($token, $session['organizationId'] ?? null)
            ->submitReview([
                'rating' => $validated['rating'],
                'body' => $validated['body'],
            ]));

        return back()->with(
            'review_status',
            $result ? 'Thanks for the review.' : 'Could not submit that review.'
        );
    }

    /** @return array<string, mixed>|null */
    private function session(Request $request): ?array
    {
        $auth = Auth::fromConfig();
        if (! $auth) {
            return null;
        }

        $cookie = $request->cookie(config('bd.session_cookie'));

        return $auth->session(is_string($cookie) ? $cookie : null);
    }
}

File: Phoenix/lib/bd_starter_web/controllers/portal_controller.ex
defmodule BdStarterWeb.PortalController do
  use BdStarterWeb, :controller

  alias BdStarter.Bd
  alias BdStarter.Bd.Auth

  @doc """
  Customer portal. Everything here is per-customer, so nothing is cached — a
  cached portal read would serve one customer's jobs to the next.
  """
  def index(conn, _params) do
    case session(conn) do
      nil ->
        redirect(conn, to: "/api/bd-auth/sign-in?returnTo=/my-account")

      session ->
        render(conn, :index,
          user: Map.get(session, "user"),
          work: Bd.portal_work(token(conn), Map.get(session, "organizationId"))
        )
    end
  end

  def submit_review(conn, params) do
    rating = params |> Map.get("rating", "5") |> to_int(5) |> min(5) |> max(1)
    body = params |> Map.get("body", "") |> String.trim()

    case session(conn) do
      nil ->
        redirect(conn, to: "/api/bd-auth/sign-in?returnTo=/my-account")

      _session when body == "" ->
        conn |> put_flash(:error, "A review needs some text.") |> redirect(to: ~p"/my-account")

      session ->
        result =
          Bd.portal_review(token(conn), Map.get(session, "organizationId"), %{
            rating: rating,
            body: body
          })

        conn
        |> put_flash(
          if(result, do: :info, else: :error),
          if(result, do: "Thanks for the review.", else: "Could not submit that review.")
        )
        |> redirect(to: ~p"/my-account")
    end
  end

  defp session(conn) do
    case Bd.client() do
      nil -> nil
      client -> Auth.session(client, token(conn))
    end
  end

  defp token(conn) do
    conn = fetch_cookies(conn)
    cookie = Application.get_env(:bd_starter, :bd_session_cookie, "bd_session")

    case conn.cookies[cookie] do
      value when is_binary(value) and value != "" -> value
      _ -> nil
    end
  end

  defp to_int(value, default) do
    case Integer.parse(to_string(value)) do
      {n, _} -> n
      _ -> default
    end
  end
end
